indigo/analysis/refs/base.py

```diff
@@ -123,3 +123,150 @@ class RefsFinderENG(BaseRefsFinder):
             )
         ''', re.X)
     candidate_xpath = ".//text()[contains(., 'Act') and not(ancestor::a:ref)]"
+
+
+class BaseSectionRefsFinder(LocaleBasedMatcher):
+    """ Finds internal references to sections in documents.
+
+    """
+
+    section_re = None
+    """ This must be defined by a subclass. It should be a compiled regular
+    expression, with named captures for `ref` and `num`.
+    """
+    # TODO: add subsection reference
+    candidate_xpath = None  # this must be defined by a subclass
+
+    # the ancestor elements that can contain references
+    ancestors = ['body', 'mainBody', 'conclusions']
+    # TODO: check on ancestor elements
+
+    def find_references_in_document(self, document):
+        """ Find references in +document+, which is an Indigo Document object.
+        """
+        # we need to use etree, not objectify, so we can't use document.doc.root,
+        # we have to re-parse it
+        root = etree.fromstring(document.content)
+        self.setup(root)
+        self.find_references(root)
+        document.content = etree.tostring(root, encoding='utf-8').decode('utf-8')
+
+    def setup(self, root):
+        self.ns = root.nsmap[None]
+        self.nsmap = {'a': self.ns}
+        self.ref_tag = f'{{{self.ns}}}ref'
+        self.ancestor_xpath = etree.XPath('|'.join(f'.//a:{a}' for a in self.ancestors), namespaces=self.nsmap)
+        self.candidate_xpath = etree.XPath(self.candidate_xpath, namespaces=self.nsmap)
+
+    def make_href(self, node, match):
+        """ Turn this match into an href if a reference is found
+        """
+        num = match.group('num')
+        candidate_elements = node.xpath(f"//a:section[a:num[text()='{num}.']]", namespaces=self.nsmap)
+        if candidate_elements:
+            element_id = candidate_elements[0].get('id')
+            return f'#{element_id}'
+
+    def find_references(self, root):
+        for ancestor in self.ancestor_nodes(root):
+            for candidate in self.candidate_nodes(ancestor):
+                node = candidate.getparent()
+
+                if not candidate.is_tail:
+                    # text directly inside a node
+                    matches = self.section_re.finditer(node.text)
+                    # mark the reference and continue to check the new tail
+                    node = self.mark_refs_in_matches(matches, node, in_tail=False)
+
+                while node is not None and node.tail:
+                    matches = list(self.section_re.finditer(node.tail))
+
+                    # if nothing in the tail matches anymore, we're done
+                    if not matches:
+                        break
+
+                    # there might be just one match that's a baddie,
+                    # in which case we're also done
+                    if len(matches) == 1 and not self.is_valid(matches[0], node):
```

Comment:
This smells weird to me. It feels like a specific instance of a general edge case: if there's nothing to do, stop. We should be able to handle it in a generic way. eg. if there are two non-valid matches in a row.